[BUG] org.opensearch.remotestore.SegmentReplicationRemoteStoreIT.testReplicaHasDiffFilesThanPrimary is flaky

#### Describe the bug
org.opensearch.remotestore.SegmentReplicationRemoteStoreIT.testReplicaHasDiffFilesThanPrimary is flaky
```
  2> java.lang.AssertionError: timed out waiting for green state
        at org.junit.Assert.fail(Assert.java:89)
        at org.opensearch.test.OpenSearchIntegTestCase.ensureColor(OpenSearchIntegTestCase.java:1002)
        at org.opensearch.test.OpenSearchIntegTestCase.ensureGreen(OpenSearchIntegTestCase.java:933)
        at org.opensearch.test.OpenSearchIntegTestCase.ensureGreen(OpenSearchIntegTestCase.java:922)
        at org.opensearch.indices.replication.SegmentReplicationIT.testReplicaHasDiffFilesThanPrimary(SegmentReplicationIT.java:780)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
        at java.base/java.lang.reflect.Method.invoke(Method.java:578)
        at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1750)
        at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:938)
        at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:974)
        at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:988)
        at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
        at org.junit.rules.RunRules.evaluate(RunRules.java:20)
.
.
.
    com.carrotsearch.randomizedtesting.UncaughtExceptionError: Captured an uncaught exception in thread: Thread[id=84, name=opensearch[node_t3][generic][T#1], state=RUNNABLE, group=TGRP-SegmentReplicationRemoteStoreIT]

        Caused by:
        java.lang.AssertionError: file (name [segment_infos_snapshot_filename__3], reused [false], length [395], recovered [477])
            at __randomizedtesting.SeedInfo.seed([E05CF3CB304D9302]:0)
            at org.opensearch.index.shard.StoreRecovery$StatsDirectoryWrapper.copyFrom(StoreRecovery.java:316)
            at org.opensearch.index.shard.IndexShard.syncSegmentsFromRemoteSegmentStore(IndexShard.java:4520)
            at org.opensearch.indices.recovery.PeerRecoveryTargetService.doRecovery(PeerRecoveryTargetService.java:248)
            at org.opensearch.indices.recovery.PeerRecoveryTargetService$RecoveryRunner.doRun(PeerRecoveryTargetService.java:604)
            at org.opensearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:806)
            at org.opensearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:52)
            at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
            at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)                                                                                      at java.base/java.lang.Thread.run(Thread.java:1589)
```

#### To Reproduce
```
./gradlew ':server:internalClusterTest' --tests "org.opensearch.remotestore.SegmentReplicationRemoteStoreIT.testReplicaHasDiffFilesThanPrimary" -Dtests.seed=E05CF3CB304D9302 -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=en-CA -Dtests.timezone=Asia/Kolkata -Druntime.java=19
```

#### Additional Context
https://build.ci.opensearch.org/job/gradle-check/15752/consoleFull

Hi @sachinpkale,
I have the `org.opensearch.remotestore.SegmentReplicationUsingRemoteStoreIT.testDeleteOperations` and `org.opensearch.remotestore.SegmentReplicationUsingRemoteStoreIT.testUpdateOperations` tests failing in [this build](https://build.ci.opensearch.org/job/gradle-check/17713/#showFailuresLink). Error is:

```
java.lang.AssertionError: Expected no missing or different segments between primary and replica but diff was missing: [name [_1_1_Lucene90_0.dvm], length [160], checksum [webb4a], writtenBy [9.7.0], name [_2.si], length [325], checksum [1auy4un], writtenBy [9.7.0], name [_2.cfe], length [479], checksum [1rwi2tu], writtenBy [9.7.0], name [_2.cfs], length [5447], checksum [cjpdu8], writtenBy [9.7.0], name [_1_1_Lucene90_0.dvd], length [87], checksum [1cgqi8w], writtenBy [9.7.0], name [_1_1.fnm], length [1216], checksum [1xrgj1t], writtenBy [9.7.0]] Different: [] Primary Replication Checkpoint : ReplicationCheckpoint{shardId=[test-idx-1][0], primaryTerm=1, segmentsGen=3, version=14, size=60012, codec=Lucene95} Replica Replication Checkpoint: ReplicationCheckpoint{shardId=[test-idx-1][0], primaryTerm=1, segmentsGen=5, version=11, size=52528, codec=Lucene95}
```

I am wondering if these 2 failures and this issue are related. They appear to map to the same base (`org.opensearch.indices.replication.SegmentReplicationBaseIT#verifyStoreContent`):

https://github.com/opensearch-project/OpenSearch/blob/0b775e79f637c7225ae908e368f1a62ce9ece011/server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationBaseIT.java#L164-L179

--- a/server/src/main/java/org/opensearch/index/shard/IndexShard.java
+++ b/server/src/main/java/org/opensearch/index/shard/IndexShard.java
@@ -4809,6 +4809,14 @@ public class IndexShard extends AbstractIndexShardComponent implements IndicesCl
                 return true;
             } else {
                 logger.warn("Checksum mismatch between local and remote segment file: {}, will override local file", file);
+                // If there is a checksum mismatch and we are not serving reads it is safe to go ahead and delete the file now.
+                // Outside of engine resets this method will be invoked during recovery so this is safe.
+                if (isReadAllowed() == false) {
+                    localDirectory.deleteFile(file);
+                } else {
+                    // segment conflict with remote store while the shard is serving reads.
+                    failShard("Local copy of segment " + file + " has a different checksum than the version in remote store", null);
+                }
             }
         } catch (NoSuchFileException | FileNotFoundException e) {
             logger.debug("File {} does not exist in local FS, downloading from remote store", file);

--- a/server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java
+++ b/server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java
@@ -457,11 +457,10 @@ public final class RemoteStoreRefreshListener extends CloseableRetryableRefreshL
     private void updateFinalStatusInSegmentTracker(boolean uploadStatus, long bytesBeforeUpload, long startTimeInNS) {
         if (uploadStatus) {
             long bytesUploaded = segmentTracker.getUploadBytesSucceeded() - bytesBeforeUpload;
-            long timeTakenInMS = (System.nanoTime() - startTimeInNS) / 1_000_000L;
-
+            long timeTakenInMS = TimeValue.nsecToMSec(System.nanoTime() - startTimeInNS);
             segmentTracker.incrementTotalUploadsSucceeded();
             segmentTracker.addUploadBytes(bytesUploaded);
-            segmentTracker.addUploadBytesPerSec((bytesUploaded * 1_000L) / timeTakenInMS);
+            segmentTracker.addUploadBytesPerSec((bytesUploaded * 1_000L) / Math.max(1, timeTakenInMS));
             segmentTracker.addUploadTimeMs(timeTakenInMS);
         } else {
             segmentTracker.incrementTotalUploadsFailed();

